server/pkcs11.go

```diff
@@ -0,0 +1,165 @@
+package server
+
+import (
+	"fmt"
+	"crypto"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cbroglie/crypto11"
+	"github.com/cloudflare/cfssl/log"
+)
+
+// PKCS11URI defines a PKCS#11 URI: https://tools.ietf.org/html/rfc7512#section-2.3
+type PKCS11URI struct { // pkcs11:path-attr[?query-attr]
+	// path-attr: (; delimited)
+	Token  string //        token <- CK_TOKEN_INFO
+	Manuf  string // manufacturer <- CK_TOKEN_INFO
+	Serial string //       serial <- CK_TOKEN_INTO
+	Model  string //        model <- CK_TOKEN_INFO
+
+	LibManuf string // library-manufacturer <- CK_INFO
+	LibDesc  string //  library-description <- CK_INFO
+	LibVer   string //      library-version <- CK_INFO
+
+	Object []byte // object <- CKA_LABEL
+	Type   string //   type <- CKA_CLASS (cert, data, private, public, or secret-key)
+	ID     []byte //     id <- CKA_ID
+
+	SlotManuf string // slot-manufacturer <- CK_SLOT_INFO
+	SlotDesc  string //  slot-description <- CK_SLOT_INFO
+	SlotID      uint //           slot-id <- CK_SLOT_ID
+
+	// query-attr: (& delimited)
+	PinSource string // pin-source
+	PinValue  string //  pin-value
+
+	ModuleName string // module-name
+	ModulePath string // module-path
+
+	// Vendor specific query attributes:
+	MaxSessions int
+}
+
+func RFC7512Parser(uri string) (*PKCS11URI, error) {
```

Comment:
My brain is shutting down trying to read the code below, please try to make it more readable. One approach: Break the function into chunks of 10-20 lines that accomplish a single coherent task, and put an inline comment above each chunk saying what that task is. Maybe just saying what some key variables represent and a brief description of the format is enough.

Reply:
Because I specifically want people to go read the RFC before reading this code. My brain also shuts down when reading the RFC. I'm sorry but it's not my fault that the authors decided to go with a scheme that has no parser anywhere. See for yourself: https://tools.ietf.org/html/rfc7512#section-2.3

I would add those comments but even I don't know what half of those variables represent!